packages/server-reason-react-ppx/server_reason_react_ppx.ml

```diff
@@ -847,7 +849,224 @@ let rewrite_jsx =
+    | _ -> Error (loc, "server-reason-react: server function arguments must have type annotations")
+
+  let response_to_json ~loc expr response =
+    let rec aux expr acc =
+      match expr.pexp_desc with
+      | Pexp_fun (_, _, _, body) -> aux body acc
+      | Pexp_constraint (expr, core_type) -> aux expr (Some core_type)
+      | _ -> (
+          match acc with
+          | Some [%type: [%t? core_type] Js.Promise.t] ->
+              let json = [%expr [%to_json: [%t core_type]] [%e response]] in
+              [%expr React.Json [%e json]]
+          | Some _ -> [%expr [%ocaml.error "server-reason-react: server functions must return a promise"]]
+          | _ ->
+              [%expr
+                [%ocaml.error "server-reason-react: server functions must have a return type annotation (Js.Promise.t)"]]
+          )
+    in
+    aux expr None
+
+  let map_arguments_to_expressions ~loc _name args =
+    List.map
+      ~f:(fun arg ->
+        match arg_details arg with
+        | Ok (arg_label, _, [%type: unit]) -> (arg_label, [%expr ()])
+        | Ok (arg_label, Some arg_name, _) -> (arg_label, [%expr [%e evar ~loc arg_name]])
+        | Ok _ ->
+            (Nolabel, [%expr [%ocaml.error "server-reason-react: invalid argument, it must have a type annotation"]])
+        | Error (loc, msg) -> (Nolabel, [%expr [%ocaml.error [%e estring ~loc msg]]]))
+      args
+
+  let encode_response ~loc name apply_args make_response =
+    [%expr
+      try
+        [%e pexp_apply ~loc [%expr [%e evar ~loc name].call] apply_args]
+        |> Lwt.map (fun response -> [%e make_response [%expr response]])
+      with e -> Lwt.fail e]
+
+  let decode_arguments_vb ~loc args args_to_decode =
+    let rec decoded_args args =
+      match args with
+      | [] -> [%expr [%e pexp_tuple ~loc (List.map ~f:(fun (_, label, _) -> evar ~loc label) args_to_decode)]]
+      | arg :: args -> (
+          match arg_details arg with
+          | Ok (_, Some label, core_type) ->
+              let of_json = make_of_json ~loc core_type [%expr [%e evar ~loc label]] in
+              [%expr
+                let [%p ppat_var ~loc { txt = label; loc }] = [%e of_json] in
+                [%e decoded_args args]]
+          | Ok _ -> decoded_args args
+          | Error _ -> decoded_args args)
+    in
+    match args_to_decode with
+    | [] -> None
+    | _ ->
+        Some
+          (value_binding ~loc
+             ~pat:[%pat? [%p ppat_tuple ~loc (List.map ~f:(fun (_, label, _) -> pvar ~loc label) args_to_decode)]]
+             ~expr:
+               [%expr
+                 match args with
+                 | [%p plist ~loc (List.map ~f:(fun (_, label, _) -> pvar ~loc label) args_to_decode)] ->
+                     [%e decoded_args args]
+                 | _ -> failwith "server-reason-react: client components need type annotations"])
+
+  let create_router_handler ~loc name args make_response =
```

Comment:
This make_response is complex and does not mean too much. Please change it to response_core_type and handle it internally